### PR TITLE
✨ Sort studies by last active

### LINKED ISSUE #520: Sort studies by Last Active
The `allStudies` query should have a parameter that allows `orderBy: "last_active"`.

This should sort and return studies based on the time of the last created version.

### PR DESCRIPTION
**Pull request by gsantia**
Closes #520 

Add a new parameter to the `allStudies` query to allow for sorting based on the time of the last created Version.

### CODE REVIEW

File: creator/settings/development.py [hunk 1]
```diff
@@ -68,6 +67,7 @@
     "creator",
     "corsheaders",
     'creator.ingest_runs.apps.IngestRunsConfig',
+    "creator.studies.apps.StudiesConfig",
```
Comment (dankolbman) [inline creator/settings/development.py:70]: You should be able to use `creator.studies` here.

**Review by dankolbman — COMMENTED** [review]
This adds a field to a study that gets updated whenever the study gets saved. The issue here is that users want to be able to see sort studies by their last active date, which is usually determined by when a study last had a document or version uploaded. The documents are a different model with a foreign key to the study so when they get created, they don't actually update the study's row.

I believe this will need to be implemented using a custom ordering filter which sorts studies by the date of latest version created in each study. This won't be as easy because there is no materialized field on the study model for this and so the query to get this ordering is going to be a bit more complex. 
To do it this way, it looks like you need to override some of the internals of graphene like is being done in this example: https://stackoverflow.com/a/61543302
graphene has a handy way to write [custom filter sets](https://docs.graphene-python.org/projects/django/en/latest/filtering/#custom-filtersets), but annoyingly, it's not so easy with ordering filters.

If we want to keep the sorting simple, you could go with the materialized field like you have here and also consider updating it whenever a new version gets added to a study. This allows a simple order by query, but it will fragment the logic for how the last active date is determined. This is problematic because if we ever want to change that logic, maybe by also considering the last action in a review or a release for that study, we will need to plug in the same update snippet in each of the places where those things happen. This would open up the possibility of us easily losing track of all the possible ways a study can determine whether it was active or not.

**Comment by dankolbman** [comment]
Before looking into how to do custom orderings, maybe try to see if adding something like this will work:
```
 order_by = OrderingFilter(
        fields=(
            ('versions__created_at',),
        )
    )
```

**Comment by dankolbman** [comment]
Something else worth exploring may be using the custom ordering filter in django-filters, which is what graphene is using under the hood: https://django-filter.readthedocs.io/en/stable/ref/filters.html?highlight=ordering#orderingfilter

File: creator/settings/development.py (hunk 1 above)
Comment (gsantia) [reply creator/settings/development.py:70]: The signal I created doesn't work using `creator.studies`

**Comment by gsantia** [comment]
> If we want to keep the sorting simple, you could go with the materialized field like you have here and also consider updating it whenever a new version gets added to a study.

I'm confused. Is this different from what I did?

**Comment by gsantia** [comment]
> Before looking into how to do custom orderings, maybe try to see if adding something like this will work:
> 
> ```
>  order_by = OrderingFilter(
>         fields=(
>             ('versions__created_at',),
>         )
>     )
> ```

`Study` has no `versions` field, right?

**Comment by dankolbman** [comment]
Ah I see that is how you did it. I commented before those signals got pushed.

Do you think you could investigate using a more advanced sorting method rather than updating the row every time something happens that we want to consider an 'active' event? If it's easy, it may be preferable for the other stated reasons but if it's not, this approach will probably be ok.

**Comment by gsantia** [comment]
> Before looking into how to do custom orderings, maybe try to see if adding something like this will work:
> 
> ```
>  order_by = OrderingFilter(
>         fields=(
>             ('versions__created_at',),
>         )
>     )
> ```

Doesn't this require a `versions` field on `Study`? Then I guess we'd need a `post_save` signal on `Version` that would add the versions as they get created to the appropriate `Study.versions`?

**Comment by gsantia** [comment]
So I've been reading into all the links provided and elsewhere, and I've been unable to come up with a query that I'd consider easy. Maybe I'm missing something. But I think we can go with the method that I have here

**Comment by gsantia** [comment]
There seems to be an issue with one of the migrations from the organizations update?

File: creator/settings/development.py (hunk 1 above)
Comment (znatty22) [reply creator/settings/development.py:71]: @gsantia if you add something like this https://github.com/kids-first/kf-api-study-creator/blob/89176bd16e9fe2ea2cefdd8eb049b41b8742264e/creator/data_templates/__init__.py into your `apps/__init__.py` you can just have `creator.studies` in your settings files

**Comment by znatty22** [comment]
@gsantia Curious to know if this works well, did you try https://github.com/kids-first/kf-api-study-creator/pull/681#issuecomment-865098170?

**Comment by gsantia** [comment]
> @gsantia Curious to know if this works well, did you try [#681 (comment)](https://github.com/kids-first/kf-api-study-creator/pull/681#issuecomment-865098170)?

I'm a bit confused by this. Should I create a new `Study.versions` field and populate it?

**Comment by znatty22** [comment]
Oh I'm sorry I forgot that studies don't have versions they have files..nvmind

**Comment by dankolbman** [comment]
@znatty22 @gsantia I believe versions do have relations to studies these days.

**Comment by gsantia** [comment]
> @znatty22 @gsantia I believe versions do have relations to studies these days.

You're right, there's a foreign key. Do you know how I should form the query?
I know that `created_at` in Graphene gets mapped to `createdAt`, but I don't know how to handle the `versions__` part.
I haven't been able to find any info online and have tried a bunch of permutations. This is what I mean:

this is the filter:
```
class StudyFilter(FilterSet):
    order_by = OrderingFilter(
        fields=(
            "name", "created_at", "organization", "active_at", "versions__created_at"
        )
    )

    class Meta:
        model = Study
        fields = ["name", "organization"]
```

and this is the GraphQL query:
```
query {
    allStudies (orderBy: "versions_createdAt") {
     edges {
       node {
         id
       }
     }
  }
}
```

Most of my attempts give:
```
"message": "['{\"order_by\": [{\"message\": \"Select a valid choice. versions_created_at is not one of the available choices.\", \"code\": \"invalid_choice\"}]}']"
```

**Comment by gsantia** [comment]
> @znatty22 @gsantia I believe versions do have relations to studies these days.

Hey... It works!

File: creator/studies/nodes.py [hunk 2]
```diff
@@ -31,6 +32,12 @@ class StudyNode(DjangoObjectType):
         description="List all releases for the study",
     )
 
+    versions = DjangoFilterConnectionField(
```
Comment (dankolbman) [inline creator/studies/nodes.py:35]: Why was this needed? Did it not allow you to sort by version properties without it?

**Review by dankolbman — COMMENTED** [review]
I don't really like how obscure `orderBy: "-versions__createdAt"` is but the `orderBy` field is sort of naturally obscure anyway since it doesn't convert things to camel and doesn't really document what's allowed in the schema.

File: creator/studies/nodes.py (hunk 2 above)
Comment (gsantia) [reply creator/studies/nodes.py:35]: yeah exactly

**Review by dankolbman — APPROVED** [review]
# 🧨
